public/app/features/alerting/unified/mocks/server/handlers/alertRules.ts

```diff
@@ -32,12 +34,41 @@ export const rulerRuleNamespaceHandler = () => {
       [namespaceByUid[folderUid].name]: namespaces[folderUid],
     });
   });
-};
 
-export const rulerRuleGroupHandler = () => {
+export const updateRulerRuleNamespaceHandler = (options?: HandlerOptions) =>
+  http.post<{ folderUid: string }>(`/api/ruler/grafana/api/v1/rules/:folderUid`, async ({ params }) => {
+    const { folderUid } = params;
+
+    // @TODO make this more generic so we can use this in other endpoints too
+    if (options?.delay) {
+      await delay(options.delay ?? 0);
```

Comment:
I checked, and it turns out it doesn't need to explicitly be some milliseconds delay - just `await delay()` is enough to ensure that the loading states are stepped through slowly enough. 
I think that without _any_ delay, MSW is performing everything all in one tick, so you end up with a instantaneous response. But `delay()` is enough for this to not happen

The creator of MSW also [recommends a pattern of using a deferred promise inside the handler here](https://github.com/mswjs/msw/discussions/1109#discussioncomment-5480488) (example is using MSW 1 though). I think the delay is fine though! Just sharing my understanding of this

Reply:
Interesting, that way the test has control over when to "resume" the handler response – it probably would also work with the new [Promise.withResolvers](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/withResolvers) instead of an external dependency :D

Reply:
So trying this with a delay set to `0` does fail the current test suite stating that "loading" is never found. This makes me think that we're either not testing it properly or arbitrary delays are still required for testing loading states :(